src/librustdoc/formats/renderer.rs

```diff
@@ -47,6 +71,40 @@ pub(crate) trait FormatRenderer<'tcx>: Sized {
     fn cache(&self) -> &Cache;
 }
 
+fn run_format_inner<'tcx, T: FormatRenderer<'tcx>>(
+    cx: &mut T,
+    item: clean::Item,
+    prof: &SelfProfilerRef,
+) -> Result<(), Error> {
+    if item.is_mod() && T::RUN_ON_MODULE {
+        // modules are special because they add a namespace. We also need to
+        // recurse into the items of the module as well.
+        let _timer =
+            prof.generic_activity_with_arg("render_mod_item", item.name.unwrap().to_string());
+
+        cx.mod_item_in(&item)?;
+        let (clean::StrippedItem(box clean::ModuleItem(module)) | clean::ModuleItem(module)) =
+            item.inner.kind
+        else {
+            unreachable!()
+        };
+        for it in module.items {
+            let info = cx.save_module_data();
+            run_format_inner(cx, it, prof)?;
```

Comment:
The old version used a heap-allocated stack, and this version instead uses the call stack.

I think that's fine, as long as you call [stacker](https://docs.rs/stacker/latest/stacker/)?

Reply:
Considering we recurse on modules (other items, they stop at one level), I think we're on the safe side. Do you see a case where it could be an issue?

Reply:
I don't know if it's going to be a problem. I just noticed that the old didn't do this.

If you don't think it's a problem, then we can go ahead with it. There might be other spots where rustdoc directly recurses on modules anyway.